nemo/collections/tts/modules/transformer_2412.py

```diff
@@ -0,0 +1,704 @@
+                self.norm_xattn_memory = torch.nn.LayerNorm(xa_d_memory, bias=False)
+
+        self.norm_pos_ff = torch.nn.LayerNorm(d_model, bias=False)
+        self.pos_ff = PositionwiseConvFF(
+            d_model, d_ffn, p_dropout, kernel_size=kernel_size, is_causal=is_causal, non_linearity=conv_non_linearity
+        )
+
+        self.use_cache = False
+        self.cache = self._init_cache()
+
+    @staticmethod
+    def _init_cache() -> Dict:
+        return {
+            'self_attn_output': None,
+            'cross_attn_output': None,
+            'memory': None,
+        }
+
+    def reset_cache(self, use_cache=False):
+        self.use_cache = use_cache
+        self.cache = self._init_cache()
+        self.self_attention.reset_cache(use_cache)
+        if self.has_xattn:
+            self.cross_attention.reset_cache(use_cache)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        x_mask: torch.Tensor,
+        cond: Optional[torch.Tensor] = None,
+        cond_mask: Optional[torch.Tensor] = None,
+        attn_prior: Optional[torch.Tensor] = None,
+    ) -> Dict:
+        """
+        Args:
+            x <torch tensor> (B, T1, C): Input tensor
+            x_mask <bool mask> (B, T1): True where ignoring is required
+            cond <torch tensor> (B, T2, C): Conditioning tensor
+            cond_mask <bool mask> (B, T2): True where ignoring is required
+
+        Returns dict with keys
+            output <torch tensor> (B, T1, C): Output tensor
+            attn_probabilities <dict>: Attention probabilities
+        """
+        x_mask_inv_float = (~x_mask).to(x.dtype).unsqueeze(-1)
+        s_attn_prob = None
+        x_, s_attn_prob = self.self_attention(query=self.norm_self(x), query_mask=x_mask)
+        if self.use_cache:
+            if self.cache['self_attn_output'] is not None:
+                x_ = torch.cat([self.cache['self_attn_output'], x_], dim=1)
+            self.cache['self_attn_output'] = x_
+        x = x + x_
+        x = x * x_mask_inv_float
+
+        x_attn_prob = None
+        if self.has_xattn and cond is not None:
+            x_normed = self.norm_xattn_query(x)
+            if self.use_cache and self.cache['memory'] is not None:
+                memory = self.cache['memory']
+            else:
+                memory = self.norm_xattn_memory(cond) if self.apply_norm_to_cond else cond
+                if self.use_cache:
+                    self.cache['memory'] = memory
+
+            x_res, x_attn_prob = self.cross_attention(
+                query=x_normed, query_mask=x_mask, memory=memory, memory_mask=cond_mask, attn_prior=attn_prior
+            )
+            if self.use_cache:
+                if self.cache['cross_attn_output'] is not None:
+                    x_res = torch.cat([self.cache['cross_attn_output'], x_res], dim=1)
+                self.cache['cross_attn_output'] = x_res
+            x = x + x_res
+            x = x * x_mask_inv_float
```

Comment:
same as above suggestion,
```suggestion
            x = (x + x_res) * x_mask_inv_float
```